### PR TITLE
Introduce stricter checks

### PR DESCRIPTION
**Pull request by DieracDelta**
Refactor testing framework. Changes include:
- Replace `DetailedTestDescription` and `GeneralTestDescription` with a control flow: `TestBuilder -> TestLauncher -> TestRunner`. Each step has clear, distinct responsibility.
- Add builders for check and setup functions that perform configurable checks
- Introduce the idea of a `Round`, which replaces the vecs of setup and checks
- Add hooks 
- Remove `PreSafetyCheck` since this can be replicated via hooks
- Reorganize errors

Remaining TODOs:
- [x] Rebase on top of HEAD of `main`
- [x] Rebase HotShot macros
- [x] Rethink one test that is commented out
- [x] Have normal tests pass
- [x] Have macro tests pass

Closes: #1068, #1097

### CODE REVIEW

File: testing/README.md
Comment (elliedavidson) [inline testing/README.md:1]: Very nice README!

**Review by elliedavidson — COMMENTED** [review]
This is absolutely wonderful!  It will be so helpful!  I don't have any requested changes.

File: consensus/src/leader.rs [hunk 1]
```diff
@@ -117,7 +117,7 @@ where
         async_sleep(self.api.propose_min_round_time() - passed_time).await;
 
         let receiver = self.transactions.subscribe().await;
-        let mut block = starting_state.next_block();
+        let mut block = <TYPES as NodeType>::StateType::next_block(Some(starting_state.clone()));
```
Comment (nyospe) [inline consensus/src/leader.rs:120]: Wow, that is an ugly clone...

File: types/src/traits/node_implementation.rs [hunk 2]
```diff
@@ -167,8 +167,6 @@ where
     I::Storage: TestableStorage<TYPES, I::Leaf>,
     TYPES::SignatureKey: TestableSignatureKey,
     I::Leaf: TestableLeaf<NodeType = TYPES>,
-    // <I::QuorumExchange as ConsensusExchange<TYPES, I::Leaf, Message<TYPES, I>>>::Membership : TestableElection<TYPES>,
```
Comment (bfish713) [inline types/src/traits/node_implementation.rs:170]: Nit: delete

**Review by bfish713 — APPROVED** [review]

File: libp2p-networking/src/network/mod.rs [hunk 3]
```diff
@@ -235,59 +235,6 @@ pub async fn gen_transport(
         .timeout(std::time::Duration::from_secs(20))
         .boxed())
 }
-// pub async fn development_transport(
```
Comment (bfish713) [inline libp2p-networking/src/network/mod.rs:238]: nit: Can we remove this

Comment (DieracDelta) [reply libp2p-networking/src/network/mod.rs:238]: Remove the commented code here? +1 that should absolutely be deleted. But I think I've deleted it?

**Pull request merged**